File: ports-py/json-rfc.py
import ports
from pathlib import Path
import json

suite = ports.suite("suites/json-rfc.ports")

@suite.placeholder("list-json-test-files")
def listTestFiles(env):
    return list(map(lambda p: p.name, Path("suites/json-rfc-fixtures").glob("*.json")))

@suite.placeholder("parse")
def parse(env,json_string):
    try:
        return json.loads(json_string)
    except Exception as e:
        return e

@suite.placeholder("parse-success?")
def parse_success(env,parse_result):
    return not isinstance(parse_result,Exception)
    
@suite.placeholder("file-contents")
def file_contents(env,file_name):
    # We need to read in binary, as the data includes invalid characters by design
    with open(f"suites/json-rfc-fixtures/{file_name}", "rb") as f:
        return f.read()
    
#
# Running
#

suite.run(exclude=(
    # Python json accespts Infinity and NaN, although they are not allowed.
    "test_n_number_infinity.json", 
    "test_n_number_NaN.json", 
    "test_n_number_minus_infinity.json"))

File: ports-rb/json-rfc.rb
require "json"
require_relative "ports"

suite "suites/json-rfc.ports" do |config|
  config.expected_failures = ["test_n_string_escaped_emoji.json"]

  placeholder "parse" do |_env, json_string|
    JSON.parse(json_string)
  rescue JSON::ParserError => e
    e
  end

  placeholder "list-json-test-files" do |_env|
    Dir.glob("suites/json-rfc-fixtures/*.json").map { |path| File.basename(path) }
  end

  placeholder "parse-success?" do |_env, parse_result|
    !parse_result.is_a?(JSON::ParserError)
  end

  placeholder "file-contents" do |_env, filename|
    File.read(File.join("suites", "json-rfc-fixtures", filename))
  end
end
